File: forensics-btmgen/src/test/java/de/burger/forensics/plugin/scan/JavaAstScannerTest.java
// DEST: src/test/java/de/burger/forensics/plugin/scan/JavaAstScannerTest.java
package de.burger.forensics.plugin.scan;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import de.burger.forensics.plugin.scan.java.JavaAstScanner;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import org.junit.jupiter.api.Test;

class JavaAstScannerTest {

    @Test
    void emitsEventsForControlFlowConstructs() throws IOException {
        Path root = Files.createTempDirectory("java-ast");
        Path file = root.resolve("Sample.java");
        Files.writeString(file, String.join("\n",
                "package com.example;",
                "",
                "public class Sample {",
                "    public int compute(int value) {",
                "        if (value > 10) {",
                "            return value;",
                "        } else {",
                "            throw new IllegalStateException();",
                "        }",
                "    }",
                "",
                "    public int choose(int value) {",
                "        switch (value) {",
                "            case 1:",
                "                return 42;",
                "            default:",
                "                return 0;",
                "        }",
                "    }",
                "}"));

        List<ScanEvent> events = new JavaAstScanner().scan(root, List.of(), List.of());
        assertFalse(events.isEmpty(), "Expected events to be emitted");

        assertTrue(events.stream()
                .filter(event -> "if-true".equals(event.kind()))
                .anyMatch(event -> "value > 10".equals(event.conditionText())
                        && "com.example.Sample".equals(event.fqcn())));
        assertTrue(events.stream()
                .filter(event -> "if-false".equals(event.kind()))
                .anyMatch(event -> "value > 10".equals(event.conditionText())));
        assertTrue(events.stream()
                .filter(event -> "switch".equals(event.kind()))
                .anyMatch(event -> "value".equals(event.conditionText())));
        assertTrue(events.stream()
                .filter(event -> "switch-case".equals(event.kind()))
                .anyMatch(event -> "case 1".equals(event.conditionText())));
        assertTrue(events.stream()
                .filter(event -> "return".equals(event.kind()))
                .anyMatch(event -> event.conditionText() == null));
        assertTrue(events.stream()
                .anyMatch(event -> "throw".equals(event.kind())));
    }
}

File: forensics-btmgen/src/test/kotlin/de/burger/forensics/plugin/scan/JavaAstScannerTest.kt
package de.burger.forensics.plugin.scan

import de.burger.forensics.plugin.scan.java.JavaAstScanner
import java.nio.file.Files
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class JavaAstScannerTest {

    @Test
    fun `emits events for control flow constructs`() {
        val root = Files.createTempDirectory("java-ast")
        val file = root.resolve("Sample.java")
        Files.writeString(
            file,
            """
            package com.example;

            public class Sample {
                public int compute(int value) {
                    if (value > 10) {
                        return value;
                    } else {
                        throw new IllegalStateException();
                    }
                }

                public int choose(int value) {
                    switch (value) {
                        case 1:
                            return 42;
                        default:
                            return 0;
                    }
                }
            }
            """.trimIndent()
        )

        val events = JavaAstScanner().scan(root, emptyList(), emptyList())
        assertThat(events).isNotEmpty
        assertThat(events).anySatisfy { event ->
            assertThat(event.kind).isEqualTo("if-true")
            assertThat(event.conditionText).isEqualTo("value > 10")
            assertThat(event.fqcn).isEqualTo("com.example.Sample")
        }
        assertThat(events).anySatisfy { event ->
            assertThat(event.kind).isEqualTo("if-false")
            assertThat(event.conditionText).isEqualTo("value > 10")
        }
        assertThat(events).anySatisfy { event ->
            assertThat(event.kind).isEqualTo("switch")
            assertThat(event.conditionText).isEqualTo("value")
        }
        assertThat(events).anySatisfy { event ->
            assertThat(event.kind).isEqualTo("switch-case")
            assertThat(event.conditionText).isEqualTo("case 1")
        }
        assertThat(events).anySatisfy { event ->
            assertThat(event.kind).isEqualTo("return")
            assertThat(event.conditionText).isNull()
        }
        assertThat(events).anySatisfy { event ->
            assertThat(event.kind).isEqualTo("throw")
        }
    }
}
